sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubUnboundedSink.java

```diff
@@ -305,27 +324,53 @@ public void startBundle(StartBundleContext c) throws Exception {
     }
 
     @ProcessElement
+    @SuppressWarnings("ReferenceEquality")
     public void processElement(ProcessContext c) throws Exception {
-      List<OutgoingMessage> pubsubMessages = new ArrayList<>(publishBatchSize);
-      int bytes = 0;
+      // TODO(sjvanrossum): Refactor the write transform so this map can be indexed with topic +
+      // ordering key and have bundle scoped lifetime.
+      // NB: A larger, breaking refactor could make this irrelevant with a GBK on topic + ordering
+      // key (or GroupIntoBatches with configurable shard for ShardedKey?) and unify
+      // bounded/unbounded writes and static/dynamic destinations.
+      Map<@Nullable String, OutgoingData> orderingKeyBatches = new HashMap<>();
+      @Nullable String currentOrderingKey = null;
+      @Nullable OutgoingData currentBatch = null;
       for (OutgoingMessage message : c.element()) {
-        if (!pubsubMessages.isEmpty()
-            && bytes + message.getMessage().getData().size() > publishBatchBytes) {
-          // Break large (in bytes) batches into smaller.
-          // (We've already broken by batch size using the trigger below, though that may
-          // run slightly over the actual PUBLISH_BATCH_SIZE. We'll consider that ok since
-          // the hard limit from Pubsub is by bytes rather than number of messages.)
-          // BLOCKS until published.
-          publishBatch(pubsubMessages, bytes);
-          pubsubMessages.clear();
-          bytes = 0;
+        // If currentBatch is null set currentOrderingKey before entering the then clause. If
+        // currentBatch is not null and currentOrderingKey does not equal messageOrderingKey set
+        // currentOrderingKey and currentBatch before entering the then or else clause and only
+        // enter the then clause if currentBatch is null. This ensures currentBatch is initialized
+        // and contains at least one element before entering the else clause if currentOrderingKey
+        // is equal to messageOrderingKey.
+        @Nullable String messageOrderingKey = message.getMessage().getOrderingKey();
+        if ((currentBatch == null
+                && (currentOrderingKey = messageOrderingKey) == messageOrderingKey)
+            || (!Objects.equals(currentOrderingKey, messageOrderingKey)
+                && (currentBatch = orderingKeyBatches.get(currentOrderingKey = messageOrderingKey))
+                    == null)) {
```

Comment:
Pretty cool trick to set variables before entering the clause! I learned something new :)

But I wonder if there is a more readable/debuggable way to write this logic, even if it takes more lines. It took me some time to figure out what was going on here (even with the explanation in the comments). This is just a preference and the code looks good as-is though so please feel free to ignore

Reply:
I had this split up into a few sections at one point, but that ended up having quite a bit of repetition. The larger refactor described above ended up getting a different batching and flushing mechanism, which I think might be applicable here as well. I'll port that over if the refactor does not get approval.